src/cljdoc/util/sqlite_cache.clj

```diff
@@ -0,0 +1,188 @@
+
+  When `ttl` is `nil` item is not stale.
+  Otherwise `ttl` is compared with `cached_ts`."
+  [{:keys [ttl cached_ts]}]
+  (if (nil? ttl)
+    false
+    (> (- (System/currentTimeMillis)
+          (.toEpochMilli (Instant/parse cached_ts)))
+       ttl)))
+
+(defn fetch-item!
+  "Performs lookup by querying on the cache table.
+  Retuens deserialized cached item."
+  [k {:keys [db-spec key-prefix deserialize-fn table key-col value-col]}]
+  (let [query (format (:fetch query-templates) table key-col)
+        row-fn #(some-> % (get (keyword value-col)) deserialize-fn)]
+    (sql/query db-spec
+               [query key-prefix (pr-str k)]
+               {:row-fn row-fn :result-set-fn first})))
+
+(defn fetch!
+  [k {:keys [db-spec key-prefix table key-col]}]
+  (let [query (format (:fetch query-templates) table key-col)]
+    (sql/query db-spec [query key-prefix (pr-str k)] {:result-set-fn first})))
+
+(defn refresh!
+  [k v {:keys [db-spec key-prefix table key-col value-col serialize-fn ttl]}]
+  (let [query (format (:refresh query-templates) table value-col key-col)
+        value (serialize-fn @v)]
+    (sql/execute! db-spec [query (Instant/now) ttl value key-prefix (pr-str k)])))
+
+(defn cache!
+  [k v {:keys [db-spec key-prefix serialize-fn table key-col value-col ttl]}]
+  (let [query (format (:cache query-templates) table key-col value-col)
+        value (serialize-fn @v)]
+    (sql/execute! db-spec [query key-prefix (Instant/now) ttl (pr-str k) value])))
+
+(defn evict!
+  [k {:keys [db-spec key-prefix table key-col]}]
+  (let [query (format (:evict query-templates) table key-col)]
+    (sql/execute! db-spec [query key-prefix (pr-str k)])))
+
+(defn seed!
+  [{:keys [db-spec key-prefix table key-col value-col]}]
+  (let [column-specs [[:ttl "integer"]
+                      [:prefix "text"]
+                      [:cached_ts "text"]
+                      [(keyword key-col) "text"]
+                      [(keyword value-col) "text"]]
+        query (sql/create-table-ddl table
+                                    column-specs
+                                    {:conditional? true})]
+    (sql/execute! db-spec [query])))
+
+(cache/defcache SQLCache [state]
+  cache/CacheProtocol
+  (lookup [_ k]
+          (delay (fetch-item! k (:cache-spec state))))
+  (lookup [this k not-found]
+          (delay (or (deref (cache/lookup this k))) not-found))
+  (has? [_ k]
+        (let [item (fetch! k (:cache-spec state))]
+          (and (not (nil? item))
+               (not (stale? item)))))
+  (hit [_ k]
+       (SQLCache. state))
+  (miss [_ k v]
+        (let [item (fetch! k (:cache-spec state))]
+          (if (and (not (nil? item)) (stale? item))
+            (refresh! k v (:cache-spec state))
+            (cache! k v (:cache-spec state))))
+        (SQLCache. state))
+  (evict [_ k]
+         (evict! k (:cache-spec state))
+         (SQLCache. state))
+  (seed [_ base]
+        (seed! (:cache-spec base))
+        (SQLCache. base))
```

Comment:
Something is odd with indentation here, I think usually protocol method bodies are indented by two spaces:

```clj
  (seed [_ base]
    (seed! (:cache-spec base))
    (SQLCache. base))
```

I also think that `seed` maybe is not be the right place for the table creation, as I understand [the docstring](
https://github.com/clojure/core.cache/blob/8fd37130ccc574ce3c9e1de460808e7d628c1d81/src/main/clojure/clojure/core/cache.clj#L37-L40) it is more meant for putting some data into the cache which is not what's being done here. Not a big deal but felt like worth mentioning.

Reply:
Fixed indent.
Yes, fair point about the `seed` fn. To me it looks like this should be used to put initial data. But our implementation requires table creation before we seed cache with any data. Thats why I added table creation in `seed`. I not sure of any better place to put table creation logic at the moment. May be we can leave it for now and evaluate later or if you have any suggestion I can implement that?